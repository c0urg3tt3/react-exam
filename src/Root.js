import PropTypes from 'prop-types'
import React, { Component } from 'react'

import { Provider } from 'react-redux'

import App from './components/app/component'
import AppHeader from './components/app-header/component'
import AppLogo from './components/app-logo/component'
import AppTitle from './components/app-title/component'
import JediList from './components/jedi-list/component'
import JediListItem from './components/jedi-list-item/component'

import Jedi from './containers/jedi'

import './Root.css'

export default class Root extends Component {
  constructor () {
    super(...arguments)
    this.state = { error: null, hasError: false }
  }

  componentDidCatch (error) {
    this.setState({ error, hasError: true })
  }

  render() {
    const { error, hasError } = this.state

    return (hasError && error.toString()) || (
      <Provider store={this.props.store}>
        <App>
          <AppHeader>
            <AppLogo/>
            <AppTitle/>
          </AppHeader>

          <Jedi>
          {({jedies}) => (
            <JediList jedies={jedies}>
              {({jedies}) => jedies.map((jedi, index) => (
                <JediListItem key={index} jedi={jedi}/>
              ))}
            </JediList>
          )}
          </Jedi>
        </App>
      </Provider>
    )
  }
}

App.propTypes = {
  children: PropTypes.any
}

App.defaultProps = {
  children: null
}
