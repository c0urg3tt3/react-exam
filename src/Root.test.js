import React from 'react'
import { render, mount, shallow } from 'enzyme'
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'

import {
  expectAppHeader,
  expectAppLogo,
  expectAppTitle,
  expectJediListLoader,
  // expectJedi,
  expectJediError,
  expectJediForm,
  expectJediFormField,
  expectJediFormButtonSubmit,
  expectJediListEmpty,
  expectJediList,
  expectJediListTitle,
  expectJediListItem,
  expectJediQuote
} from './utils/expectedComponents'

const middlewares = [thunk]
const mockStore = configureMockStore(middlewares)

import Root from './Root'

describe('Root', () => {
  it('should render component', () => {
    const store = mockStore({ jedi: { list: { jedies: [] }}})
    const wrapper = mount(<Root store={store}/>)

    expect(wrapper.exists()).toEqual(true)
    expect(wrapper.find('.app').exists()).toEqual(true)

    expectAppHeader({wrapper})
    expectAppLogo({wrapper: wrapper.find('AppHeader')})
    expectAppTitle({wrapper})
    expectJediForm({wrapper})
    expectJediFormField({
      wrapper,
      elementId: 'jedi-form-field_name',
      labelValue: "Add a Jedi to the Council",
      placeholder: "name",
      value: ""
    })
    expectJediFormButtonSubmit({
      wrapper,
      title: "Submit your padawan to the Council",
      isDisabled: true
    })
    expectAppLogo({
      wrapper: wrapper.find('JediFormButtonSubmit'),
      isAnimate: false
    })
    expectJediList({wrapper, exist: false})
    expectJediListItem({wrapper, exist: false})
    expectJediListEmpty({wrapper})
    expectJediQuote({
      wrapper: wrapper.find('JediListEmpty'),
      message: "Looks like someone executed Order 66 !",
      quote: "Master sky walker there is too many of them what are we going to do?",
      autor: "Padawan"
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render list item when jedi fetch success', () => {
    const jedies = [
      {id: 42, name: 'Jar Jar Bings'},
      {id: 1337, name: 'Anakin Skywalkers'}
    ]
    const store = mockStore({ jedi: { list: { jedies }}})
    const wrapper = mount(<Root store={store}/>)

    expectJediList({wrapper, jedies})
    expectJediListTitle({wrapper})
    expectJediListItem({wrapper, jedi: jedies[0]})

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render list loader when jedi fetch in progress', () => {
    const store = mockStore({ jedi: { list: { isFetching: true }}})
    const wrapper = mount(<Root store={store}/>)

    expectJediListLoader({wrapper})
    expectJediQuote({
      wrapper: wrapper.find('JediListLoader'),
      message: "C3PO is fetching jedi list ...",
      quote: "hnnnhrrhhh huurh uughghhhgh raaaaaahhgh uughguughhhghghghhhgh",
      autor: "Chewbacca"
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render error when jedi fetch in failure', () => {
    const errorMessage = "death star destroyed !"
    const store = mockStore({ jedi: { list: {
      isFetchingError: true,
      fetchingError: errorMessage
    }}})
    const wrapper = mount(<Root store={store}/>)

    expectJediError({
      wrapper,
      isFetchingError: true,
      errorMessage: errorMessage
    })
    expectJediQuote({
      wrapper: wrapper.find('JediError'),
      message: errorMessage,
      quote: "What if I told you that the Republic was now under the control of a Dark Lord of the Sith?",
      autor: "Count Dooku"
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render error when jedi post in failure', () => {
    const errorMessage = "death star destroyed !"
    const store = mockStore({ jedi: { list: {
      isPostingError: true,
      postingError: errorMessage
    }}})
    const wrapper = mount(<Root store={store}/>)

    expectJediError({
      wrapper,
      isPostingError: true,
      errorMessage: errorMessage
    })
    expectJediQuote({
      wrapper: wrapper.find('JediError'),
      message: errorMessage,
      quote: "The Council will act as they deem necessary.",
      autor: "Mace Windu"
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
