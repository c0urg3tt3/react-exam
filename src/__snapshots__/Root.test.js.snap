// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Root should render component 1`] = `"<div class=\\"app\\"><header class=\\"app-header\\"><img alt=\\"app logo\\" src=\\"logo_jedi_order.svg\\" class=\\"app-logo rotate\\"><h1 class=\\"app-title\\">Jedi List</h1></header><div class=\\"jedi\\"><form class=\\"jedi-form\\"><div class=\\"jedi-form-field\\"><label class=\\"jedi-form-field-label\\" for=\\"jedi-name-input\\">Add a Jedi to the Council</label><input class=\\"jedi-form-field-input\\" id=\\"jedi-name-input\\" name=\\"jedi-name-input\\" value=\\"\\" placeholder=\\"name\\"></div><button class=\\"jedi-form-button-submit\\" disabled=\\"\\">+</button></form><div class=\\"jedi-list-empty\\"><div class=\\"jedi-quote\\"><p>Looks like someone executed Order 66 !</p><blockquote><p>Master sky walker there is too many of them what are we going to do?</p><footer>Padawan</footer></blockquote></div></div></div></div>"`;

exports[`Root should render error when jedi fetch in failure 1`] = `"<div class=\\"app\\"><header class=\\"app-header\\"><img alt=\\"app logo\\" src=\\"logo_jedi_order.svg\\" class=\\"app-logo rotate\\"><h1 class=\\"app-title\\">Jedi List</h1></header><div class=\\"jedi error\\"><div class=\\"jedi-error\\"><div class=\\"jedi-quote\\"><p>death star destroyed !</p><blockquote><p>What if I told you that the Republic was now under the control of a Dark Lord of the Sith?</p><footer>Count Dooku</footer></blockquote></div></div><form class=\\"jedi-form\\"><div class=\\"jedi-form-field\\"><label class=\\"jedi-form-field-label\\" for=\\"jedi-name-input\\">Add a Jedi to the Council</label><input class=\\"jedi-form-field-input\\" id=\\"jedi-name-input\\" name=\\"jedi-name-input\\" value=\\"\\" placeholder=\\"name\\"></div><button class=\\"jedi-form-button-submit\\" disabled=\\"\\">+</button></form><div class=\\"jedi-list-empty\\"><div class=\\"jedi-quote\\"><p>Looks like someone executed Order 66 !</p><blockquote><p>Master sky walker there is too many of them what are we going to do?</p><footer>Padawan</footer></blockquote></div></div></div></div>"`;

exports[`Root should render list item when jedi fetch success 1`] = `"<div class=\\"app\\"><header class=\\"app-header\\"><img alt=\\"app logo\\" src=\\"logo_jedi_order.svg\\" class=\\"app-logo rotate\\"><h1 class=\\"app-title\\">Jedi List</h1></header><div class=\\"jedi\\"><form class=\\"jedi-form\\"><div class=\\"jedi-form-field\\"><label class=\\"jedi-form-field-label\\" for=\\"jedi-name-input\\">Add a Jedi to the Council</label><input class=\\"jedi-form-field-input\\" id=\\"jedi-name-input\\" name=\\"jedi-name-input\\" value=\\"\\" placeholder=\\"name\\"></div><button class=\\"jedi-form-button-submit\\" disabled=\\"\\">+</button></form><ul class=\\"jedi-list\\"><li class=\\"jedi-list-item\\"><span>42</span><span>Jar Jar Bings</span></li><li class=\\"jedi-list-item\\"><span>1337</span><span>Anakin Skywalkers</span></li></ul></div></div>"`;

exports[`Root should render list loader when jedi fetch in progress 1`] = `"<div class=\\"app\\"><header class=\\"app-header\\"><img alt=\\"app logo\\" src=\\"logo_jedi_order.svg\\" class=\\"app-logo rotate\\"><h1 class=\\"app-title\\">Jedi List</h1></header><div class=\\"jedi\\"><form class=\\"jedi-form\\"><div class=\\"jedi-form-field\\"><label class=\\"jedi-form-field-label\\" for=\\"jedi-name-input\\">Add a Jedi to the Council</label><input class=\\"jedi-form-field-input\\" id=\\"jedi-name-input\\" name=\\"jedi-name-input\\" value=\\"\\" placeholder=\\"name\\"></div><button class=\\"jedi-form-button-submit\\" disabled=\\"\\">+</button></form><div class=\\"jedi-list-loader\\"><div class=\\"jedi-quote\\"><p>C3PO is fetching jedi list ...</p><blockquote><p>hnnnhrrhhh huurh uughghhhgh raaaaaahhgh uughguughhhghghghhhgh</p><footer>Chewbacca</footer></blockquote></div></div></div></div>"`;
